mypy/argmap.py

```diff
@@ -211,6 +212,18 @@ def expand_actual_type(
                     # Just return `Any`, other parts of code would raise
                     # a different error for improper use.
                     return AnyType(TypeOfAny.from_error)
+            elif isinstance(actual_type, UnionType):
+                item_types = [
+                    self.expand_actual_type(
+                        item,
+                        actual_kind=actual_kind,
+                        formal_name=formal_name,
+                        formal_kind=formal_kind,
+                        allow_unpack=allow_unpack,
+                    )
+                    for item in actual_type.items
+                ]
```

Comment:
This doesn't work because the mapper has mutable state (`self.tuple_index`), so one would need to create multiple copies of the current mapper and apply them independently.

But then this will run into problems if the union is of differently sized tuple, since then you end up with incongruent state across the different copies.

Solving this problem generally seems intractable due to combinatorial explosion: consider `f(*x1, *x2, ..., *xn)`. If each $`xₖ`$ is comprised of a union of $`mₖ`$ differently sized tuples, then there are $`m₁⋅m₂⋅…⋅mₙ`$ possible paths.

So at best I think one can do this union splitting if the Union doesn't contain tuples, or if all contained tuples are of equal length.